subsys/dfu/dfu_target/src/dfu_stream_flatten.c

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <dfu_stream_flatten.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(dfu_stream_flatten, CONFIG_DFU_TARGET_LOG_LEVEL);
+
+int stream_flash_flatten_page(struct stream_flash_ctx *ctx, off_t off)
+{
+	int rc;
+	struct flash_pages_info page;
+
+#if defined(CONFIG_STREAM_FLASH_ERASE) && defined(CONFIG_FLASH_HAS_EXPLICIT_ERASE) && \
+defined(CONFIG_FLASH_HAS_NO_EXPLICIT_ERASE)
+	bool update_page = false;
+
+	const struct flash_parameters *fparams = flash_get_parameters(ctx->fdev);
+
+	if ((flash_params_get_erase_cap(fparams) & FLASH_ERASE_C_EXPLICIT)) {
+		update_page = true;
+	}
+#endif
+
+	rc = flash_get_page_info_by_offs(ctx->fdev, off, &page);
+
+	if (rc != 0) {
+		LOG_ERR("Error %d while getting page info", rc);
+		return rc;
+	}
+
+#if defined(CONFIG_STREAM_FLASH_ERASE)
+#if defined(CONFIG_FLASH_HAS_EXPLICIT_ERASE) && defined(CONFIG_FLASH_HAS_NO_EXPLICIT_ERASE)
+	if (!update_page) {
+	} else
+#endif
+	if (ctx->last_erased_page_start_offset == page.start_offset) {
```

Comment:
need to preserve content against being flattened for RRAM-only as well

Reply:
Fixed

Reply:
image might be uploaded as complete, which mean that it consist image trailer flags (imgtool --confirm). That why the same check exist in the stream_flash_page_erase.
In this code same thing should be preserved for CONFIG_STREAM_FLASH_ERASE=n case.

I did a try to evaluate this for that case:
``` C
			ssize_t status_offset = boot_get_trailer_status_offset(secondary_size[cur_sec_img]);
			size_t offset;
			err = dfu_target_stream_offset_get(&offset);

			if (err == 0) && (page.start_offset >= offset)
			{
				flash_flatten
			}
```

Eventually we can just use flash_flatten in  stream_flash_page_erase.